src/jit/lowerxarch.cpp

```diff
@@ -978,6 +978,63 @@ void Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
             LowerHWIntrinsicCC(node, NI_AVX_PTEST, GenCondition::UGT);
             break;
 
+        case NI_FMA_MultiplyAddScalar:
+        {
+            // Math(F).FusedMultiplyAdd is expanded into NI_FMA_MultiplyAddScalar but
+            // depending on additional GT_NEG nodes it can be:
+            //
+            //    x *  y + z -> NI_FMA_MultiplyAddScalar
+            //    x * -y + z -> NI_FMA_MultiplyAddNegatedScalar
+            //   -x *  y + z -> NI_FMA_MultiplyAddNegatedScalar
+            //   -x * -y + z -> NI_FMA_MultiplyAddScalar
+            //    x *  y - z -> NI_FMA_MultiplySubtractScalar
+            //    x * -y - z -> NI_FMA_MultiplySubtractNegatedScalar
+            //   -x *  y - z -> NI_FMA_MultiplySubtractNegatedScalar
+            //   -x * -y - z -> NI_FMA_MultiplySubtractScalar
+            //
+            GenTreeArgList* argList = node->gtGetOp1()->AsArgList();
+            if (argList->Current()->OperIsHWIntrinsic()) // CreateScalarUnsafe is expected
```

Comment:
Is this assumption valid?

This is only checking for `NI_FMA_MultiplyAddScalar` and so this could be a `Fma.MultiplyAdd(LoadScalarVector128(), x, y)`, could it not?

Reply:
Same goes for the other checks below...

Reply:
Maybe it is late but I don't understand, could you please explain?

> Is this assumption valid?

Do you mean `// CreateScalarUnsafe is expected` assumption?

> This is only checking for NI_FMA_MultiplyAddScalar 

Do you mean `case NI_FMA_MultiplyAddScalar:` check?

> and so this could be a Fma.MultiplyAdd(LoadScalarVector128(), x, y), could it not?

Do you mean that the name intrinsic that reaches that code could be `NI_System_Math_FusedMultiplyAdd` or `NI_System_MathF_FusedMultiplyAdd`?

what is the issue here?

Reply:
This definitely needs an explicit check (or at least an assert if the assumption is actually valid, which I do not believe).

Reply:
> System.Runtime.Intrinsics.Fma.MultiplyAddScalar.

How do we create IR for that? I do not see this name in `importer::lookupNamedIntrinsic`.

Reply:
Hardware Intrinsics are handled more generally by `HWIntrinsicInfo::lookupId`: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L4320

That method (https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsic.cpp#L186) then matches against the hardware intrinsic list for a given platform: https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsiclistxarch.h

The `System.Runtime.Intrinsics.Fma.MultiplyAddScalar` is specifically here: https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsiclistxarch.h#L589

Hardware Intrinsics themselves are imported via `impHWIntrinsic`: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L3473 which generates `GenTreeHWIntrinsic` nodes that are handled by special codegen and importation logic.

Reply:
So, while `Math.FusedMultiplyAdd` and `MathF.FusedMultiplyAdd` will create explicit hardware intrinsic nodes matching the above shape: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L4019

`Fma.MultiplyAddScalar` can be more arbitrary and could, for example, have x/y/z be `OperIsHWIntrinsic` nodes where the intrinsicID was `NI_SSE_LoadVector128`: https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsiclistxarch.h#L127